lib/Backends/CPU/LLVMIRGen.cpp

```diff
@@ -637,17 +666,35 @@ void LLVMIRGen::generateLLVMIRForModule(llvm::IRBuilder<> &builder) {
     // This is a data parallel instruction.
 
     // Check if the current instruction is shape compatible with the bundle.
-    bool isShapeCompatible = true;
+    bool isBundleCompatible = true;
     if (!bundle.empty()) {
       auto val = I->getOperand(0).first;
       auto bundleVal = bundle.back()->getOperand(0).first;
       // Check if shapes have the same amount of elements.
-      isShapeCompatible = val->size() == bundleVal->size();
+      isBundleCompatible = val->size() == bundleVal->size();
+    }
+
+    // Check all mutated operands of the current instruction. Their memory
+    // regions should not have a non-exact overlap with any operands of the
+    // bundled instructions. In case this condition does not hold, the current
+    // instruction cannot be included into the data-parallel bundle, because
+    // overlapping operand buffers are not data parallel.
+    for (auto op : I->getOperands()) {
+      // Skip non-mutated operands.
+      if (op.second == OperandKind::In)
+        continue;
+      // If the mutated operand buffer overlaps with any buffer already used by
+      // the bundle, the current instruction cannot become a part of the bundle.
+      if (isOverlappingWithAnyBundleBufferOperands(allocationsInfo_, bundle,
+                                                   op.first)) {
+        isBundleCompatible = false;
+        break;
+      }
     }
 
     // If the instruction is not shape-compatible, emit the kernel for the
```

Comment:
can you update comment here, since it's not only shape compatibility which matters.